Adds support for implicit route model binding with translated slugs

marvin-wtt opened the pull request:
Adds implicit route model binding support for translated slugs using [spatie/laravel-translatable](https://github.com/spatie/laravel-translatable) if the route key matches the slug field

freekmurze commented:
Seems interesting, but currently tests and updated readme seem to be missing. Could you take care of that?

marvin-wtt commented:
@freekmurze Tanks for the reply. I updated the readme and added a brief explanation. Unfortunately I was unable to add test cases because sqlite doesn't support json without the [JSON1 extention](https://www.sqlite.org/json1.html). I added a try-catch block  around the added operation to avoid conflicts with databases that don't support json.  
Do you want me to squash the commits?

freekmurze commented:
In that case, we'll need to switch to using MySQL for the tests. Currently I don't have the time to take care of this, but I accept a separate PR that does this.

marvin-wtt commented:
Turns out I didn't read the [laravel docs](https://laravel.com/docs/8.x/queries#json-where-clauses) about `JSON Where Clauses` properly. Fortunately SQLite does support json but not Laravels `whereJsonContains` method. I updated it to use the the `JSON_EXTRACT()` function instead and it works for testing.

freekmurze commented on src/HasTranslatableSlug.php:
```diff
@@ -120,4 +123,63 @@ protected function hasCustomSlugBeenUsed(): bool
+                    $relationship->getRelated()->getTable() . '.' . $field . '->' . $this->getLocale(),
+                    $value
+                );
+            } else {
```
Could you refactor this `else` away?

freekmurze replied:
That else isn't necessary as you're returning right above it.

freekmurze commented on src/HasTranslatableSlug.php:
```diff
@@ -120,4 +123,63 @@ protected function hasCustomSlugBeenUsed(): bool
+            if ($relationship instanceof HasManyThrough ||
+                $relationship instanceof BelongsToMany) {
+                return $relationship->where(
+                    $relationship->getRelated()->getTable() . '.' . $field . '->' . $this->getLocale(),
```
Please use string interpolation where possible

freekmurze commented on src/HasTranslatableSlug.php:
```diff
@@ -120,4 +123,63 @@ protected function hasCustomSlugBeenUsed(): bool
+            } else {
+                return $relationship->where($field . '->' . $this->getLocale(), $value);
+            }
+        } catch (\RuntimeException $exception) {
```
Import this so we can lose `\`

freekmurze commented on tests/HasTranslatableSlugTest.php:
```diff
@@ -302,4 +302,56 @@ public function it_can_update_slug_with_non_unique_names_multiple()
+        $model->setTranslation('slug', 'nl', 'updated-value-nl');
+        $model->save();
+
+        // Test for em locale
```
Typo?

freekmurze commented on src/HasTranslatableSlug.php:
```diff
@@ -120,4 +123,63 @@ protected function hasCustomSlugBeenUsed(): bool
+
+    public function resolveRouteBinding($value, $field = null): Model|null
+    {
+        if (($field ?? $this->getRouteKeyName()) !== $this->getSlugOptions()->slugField) {
```
Could you break this up to make it more readable?

freekmurze commented on src/HasTranslatableSlug.php:
```diff
@@ -120,4 +123,63 @@ protected function hasCustomSlugBeenUsed(): bool
+        // Only some database types support json operations.
+        // If the database doesn't support it, null is returned as default method would do the same
+        try {
+            return $this->where($this->getSlugOptions()->slugField . '->' . $this->getLocale(), $value)
```
please put `->where` on a new line

freekmurze commented on tests/HasTranslatableSlugTest.php:
```diff
@@ -302,4 +302,56 @@ public function it_can_update_slug_with_non_unique_names_multiple()
+    /** @test */
+    public function it_can_resolve_route_binding()
+    {
+        $model = new TranslatableModel();
```
Instead of testing it on this low level, create a test in which you create a Route, set up the binding and check if the resolved model that gets past to the route is the model you expect. That would give me more confidence that everything is working in tandem with Laravel.

freekmurze commented on tests/TestCase.php:
```diff
@@ -64,6 +64,15 @@ protected function setUpDatabase(Application $app)
             $table->text('slug')->nullable();
         });
 
+        $app['db']->connection()->getSchemaBuilder()->create('translatable_model_soft_deletes', function (Blueprint $table) {
```
Just use `Schema::create` here

freekmurze replied:
Could you take care of this?

freekmurze commented on tests/TranslatableModelSoftDeletes.php:
```diff
@@ -0,0 +1,37 @@
+
+    public $translatable = ['name', 'other_field', 'slug'];
+
+    private $customSlugOptions;
```
Could this be `protected`?

freekmurze replied:
This should be changed too, so that it follows our conventions

freekmurze commented:
Let me know when this is ready for review.

marvin-wtt commented:
I added some questions to your change request comments where I was unsure what to do. Except that, it's ready for review.

freekmurze commented on README.md:
````diff
@@ -339,6 +339,44 @@ class YourEloquentModel extends Model
     }
 }
 ```
+You can also use Laravels [implicit route model binding](https://laravel.com/docs/8.x/routing#implicit-binding) inside your controller to automatically resolve the model. To use this feature, make sure that the slug column matches the `routeNameKey`.  
````
Add a subtitle here

freekmurze commented on src/HasTranslatableSlug.php:
```diff
@@ -120,4 +124,62 @@ protected function hasCustomSlugBeenUsed(): bool
+        }
+    }
+
+    public function resolveChildRouteBindingQuery($childType, $value, $field): Model|null
```
Add a test that makes sure that this functionality works.

freekmurze commented on src/HasTranslatableSlug.php:
```diff
@@ -120,4 +124,62 @@ protected function hasCustomSlugBeenUsed(): bool
+        }
+    }
+
+    public function resolveSoftDeletableRouteBinding($value, $field = null): Model|null
```
Isn't this nearly the same as `resolveRouteBinding`? Could we remove this duplication somehow?

marvin-wtt replied:
You are right. It is a lot of duplication. I used the original implementation and only updated what I needed to avoid any bug introduced by myself.

freekmurze commented on tests/HasTranslatableSlugTest.php:
```diff
@@ -302,4 +303,75 @@ public function it_can_update_slug_with_non_unique_names_multiple()
+        $model->save();
+
+        // Test for en locale
+        $result = (new TranslatableModel())->resolveRouteBinding('updated-value-en', 'slug');
```
Instead of calling these methods directly, you could register a route that uses model binding and check if you get the desired model. That would give more confidence that everything is working correctly in tandem with Laravel.

Routes can be registered from anywhere, so also in this very test.

marvin-wtt replied:
I did this in `#it_can_bind_route_model_implicit`

freekmurze commented on tests/TestCase.php:
```diff
@@ -26,45 +28,54 @@ protected function getEnvironmentSetUp($app)
 
-        $app['config']->set('database.default', 'sqlite');
-        $app['config']->set('database.connections.sqlite', [
+        Config::set('database.default', 'sqlite');
```
Let's use the `config()` helper instead of the facade. That's more in line with our other packages.

freekmurze reviewed (changes requested):
It's going in the right direction 👍

marvin-wtt commented:
I just realized that my comments mentioned above where still pending :see_no_evil:. I'll take care of the change request this evening

marvin-wtt commented:
@freekmurze As you mentioned in the review comment, child model resolution was not working correctly as it is only handled by the parent model. I created a PR which fixes this problem: https://github.com/laravel/framework/pull/39929 . This however means that implicit route model binding only works from the next release (probably v8.76.0+). I could also keep the current changes to at least support implicit route model binding for the parent model for lower version - this would be unnecessary for all future versions though. 
How would you prefer to proceed here?

freekmurze commented:
> As you mentioned in the review comment, child model resolution was not working correctly as it is only handled by the parent model. I created a PR which fixes this problem: laravel/framework#39929 . This however means that implicit route model binding only works from the next release (probably v8.76.0+).

Instead of checking for the specific Laravel version number (and adding a note), feel free to update the required Laravel version in composer.json 👍

Seems like the tests are failing, could you take a look at that?

marvin-wtt commented:
> Instead of checking for the specific Laravel version number (and adding a note), feel free to update the required Laravel version in composer.json 👍

I can do that for laravel framework v8 but not for v7.

>Seems like the tests are failing, could you take a look at that?

The test fails because the version I check on should be replaced with the next release. I tested with `8.x-dev` and all tests passed. I'll fix that as soon as the new framework version is released

freekmurze commented:
> I can do that for laravel framework v8 but not for v7.

Let's just drop support for Laravel 7.

>  I'll fix that as soon as the new framework version is released

Great, thanks! Feel free to poke me again then.

marvin-wtt commented:
I updated the framework version to `v8.76.1`

@freekmurze Do you think it would be useful to not only query for the current locale but for all enabled locales to better support url sharing? Currently url sharing is only possible within the same locale

freekmurze commented:
> Do you think it would be useful to not only query for the current locale but for all enabled locales to better support url sharing?

I think just enable this for the current locale is fine 👍

Thanks for your work on this PR.

marvin-wtt commented:
@freekmurze First of all thanks for merging 

> I think just enable this for the current locale is fine 👍

I need this for a project anyway. The question is if you'd like to have this as a feature in this project and what your opinion on a separate table to improve querying is. 

My idea would be like this:
```
id -> bigint
model -> string
model_id -> bigint
locale -> string
slug -> string
```
This could be a PR for laravel-translatable as well but I think it is intended to keep it as json for simplicity.

Alternatively I could also image to query the model like this:
```php
foreach ($locales as $locale) {
    $query->orWhere("{$field}->{$locale}", $value);
}
```
But afaIk, it is bad practice to query with json like this

keizah7 commented on composer.json:
```diff
@@ -17,7 +17,7 @@
     "require": {
         "php": "^8.0",
-        "illuminate/database": "^8.0",
+        "illuminate/database": "^8.76.1",
```
```
sail@f34348e1ecf9:/var/www/html$ composer require spatie/laravel-sluggable
Using version ^3.2 for spatie/laravel-sluggable
./composer.json has been updated
Running composer update spatie/laravel-sluggable
Loading composer repositories with package information
Updating dependencies
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Root composer.json requires spatie/laravel-sluggable ^3.2 -> satisfiable by spatie/laravel-sluggable[3.2.0].
    - spatie/laravel-sluggable 3.2.0 requires illuminate/database ^8.76.1 -> found illuminate/database[v8.76.1, ..., 8.x-dev] but these were not loaded, likely because it conflicts with another require.


Installation failed, reverting ./composer.json and ./composer.lock to their original content.

```

marvin-wtt replied:
What other packages do you use? And which version do you use for your project?

keizah7 replied:
![image](https://user-images.githubusercontent.com/49227202/147582094-8cd14cfb-7780-4b31-961d-e5d16cfdb06c.png)
I was able to install only after updating all packages
